src/ripple/rpc/impl/Handler.cpp

```diff
@@ -129,8 +129,8 @@ Handler const handlerArray[]{
      Role::ADMIN,
      NO_CONDITION},
     {"ripple_path_find", byRef(&doRipplePathFind), Role::USER, NO_CONDITION},
-    {"sign", byRef(&doSign), Role::USER, NO_CONDITION},
-    {"sign_for", byRef(&doSignFor), Role::USER, NO_CONDITION},
+    {"sign", byRef(&doSign), Role::USER, NEEDS_CURRENT_LEDGER},
+    {"sign_for", byRef(&doSignFor), Role::USER, NEEDS_CURRENT_LEDGER},
```

Comment:
Why? Questions about the wisdom of even allowing this command at all, requiring a current ledger basically makes it impossible to use `rippled` for offline signing on an airgapped machine, which is something that people do.

Reply:
the `sign` and `signFor` methods can auto fill certain fields, in which case the server does need access to the current ledger. I could change this condition back to 'NO_CONDITION` and try to detect when the handler is trying to auto fill fields (in which case the server would then forward the request to a p2p node).

Reply:
In fact, `RPC::transactionSign` tries to access the open ledger regardless: https://github.com/cjcobb23/rippled/blob/6abe7a57bb77342579a9d5608a8c5304adbdc214/src/ripple/rpc/impl/TransactionSign.cpp#L772
Reporting mode has no concept of the open ledger, so it needs to forward the request to a p2p node.

Reply:
I think the cleanest thing here is to change the condition back to `NO_CONDITION` (though it does seem a little weird to me that an RPC with condition `NO_CONDITION` reads from `app.openLedger().current()`), and have `app.openLedger()` throw a `ReportingShouldProxy` exception if the server is in reporting mode. Throwing this exception causes the request to be forwarded to a p2p node.

Reply:
@cjcobb23 I understand. We need this to be `NO_CONDITION` because offline signing is a hard requirement until there is another tool that can do this for users. I think that we should _only_ access `app.openLedger()` if we actually need to fill in information. If the fields are present, we should not look at the open ledger at all.

Reply:
@nbougalis I changed this back to `NO_CONDITION` in https://github.com/ripple/rippled/pull/3609/commits/02ca767df9330181c8685ca23f417361d2f4c66d. I also changed the code so we only access `app.openLedger()` if we need to fill in fields. However, we were also using the `Rules` of the open ledger, to check if a fully canonical signature is required. For reporting mode, I changed this to just use the `Rules` of the most recently validated ledger, but for non-reporting mode, we still uses the Rules of the open ledger. Let me know what you think of this.

Reply:
Actually, I slightly misspoke here. Even if no fields need to be filled in, the server still uses the open ledger to verify parts of the transaction. These checks are skipped if `offline` is set to `true` in the request. So the change I made was to only access `app.openLedger()` if `offline` is `true`. However, we still use the open ledger for the `Rules`, though reporting will just use the `Rules` from the validated ledger.